### PR TITLE
Record pool and turn metrics through the telemetry adapter

### PR DESCRIPTION
**Pull request by samuelcolvin**
Spans describe one session; they cannot answer whether the fleet is saturated, how often workers die, or how much of its duration budget a typical run uses. Add the aggregate side: five pool-health instruments (worker gauge, checkout wait, spawn cost, terminations by reason, session lifetime) and eleven per-turn ones, recorded by a `TurnMetrics` state machine that mirrors the protocol the way `telemetry::Recorder` does for spans — separate from it, because metrics must also cover untraced checkouts.

`monty.run.duration` minus `monty.run.execution_time` is time the host spent answering suspensions, which is where latency usually turns out to be. Attribute cardinality is enforced rather than hoped for: a function name is recorded only when the host resolved it, `exc_type` round-trips through `ExcType` and collapses to `other` otherwise, and no path reaches an attribute — sandboxed code must not be able to mint time series.

Measurements reach hosts two ways. A foreign SDK gets them through `TelemetryAdapter::record_metric`, which defaults to a no-op so adapters written before metrics keep working and TELEMETRY_ADAPTER_VERSION stays 1; the Python bridge probes for the method, the Node bridge emits a `metric` event kind. A Rust host uses `Metrics::for_logfire` instead and gets real instruments on its own meter, with exponential histogram buckets rather than the SDK defaults that would put every monty turn in one bucket.

The worker gauge is a gauge, not an up/down counter: one decrement missed on an error path would skew an up/down counter for the process's life. Recording never happens under the pool mutex, since it reaches into the host SDK and, for Python, its GIL.

Also run `cargo test -p monty-pool` with `telemetry-adapter` enabled in the Makefile and CI, without which none of this — nor the existing telemetry tests — is covered.

Claude-Session: https://claude.ai/code/session_0177dP9KE9Son2yHh3vgtMim

---
## Summary by cubic
Adds pool-wide and per-turn metrics to `monty-pool`, so hosts can monitor fleet saturation, worker turnover, session lifetimes, and turn latency instead of only individual sessions. Metrics now use an OpenTelemetry exporter pipeline: Rust hosts use native meters, while Python and Node adapters receive aggregated OTLP batches.

- **Metrics**
  - Records live and host-blocked workers, checkout waits, terminations by reason, session duration, execution time, suspension round trips, and turn outcomes.
  - Uses closed attribute sets so sandbox-controlled names, exception types, and paths never create metric dimensions.
  - Keeps shared worker gauges and housekeeping-turn timing accurate across races, errors, restores, and pool drops.

- **Integration**
  - Renames the feature to `telemetry`, retaining `telemetry-adapter` as an alias; bindings enable it and configure `PoolConfig.metrics` automatically.
  - Rust hosts must set `PoolConfig.metrics = Metrics::for_logfire(...)` or use `TelemetryAdapterHandle::metrics`.
  - Adapter version 1 remains compatible; optional metric export delivers OTLP batches, with `_flush_telemetry` and `_flushTelemetry` available before shutdown.

<sup>Written for commit c210a5895675d846d7b7b42b6e5c8563c344eef0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/pydantic/monty/pull/686?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by samuelcolvin** [comment]
_(Written by claude)_

## Security review

Scope: the whole branch diff, weighted toward host/parent code, with a specific look at whether the new metrics code can cause resource exhaustion. `heap.rs`, `path_security.rs`, `mount_table.rs` and the proto schema are untouched; the only wire-path change is `FrameRecv::recv` returning the frame length it had already validated against `MAX_FRAME_LEN`, so decoding and its validation are unchanged.

### The cardinality question is handled — this was the main risk and it's closed

Every attribute recorded is a closed set fixed in `metrics.rs`: `outcome`, `reason`, `turn`, `kind`, `direction`, `stream`, and `function` (from the protocol's own `os_call` oneof). Nothing the sandbox names — a called function's name, an exception class, a looked-up name, a path — reaches an attribute, under any outcome. `crates/monty-pool/src/telemetry/metrics.rs:1051` tests exactly the case that would otherwise slip through (`AttributeError` from a host object comes back as `error`, not `not_found`). Hostile values from a compromised child are also safe: `total_execution_micros: u64` goes through `saturating_sub` + `Duration::from_micros` (total), and byte lengths through saturating `i64::try_from`. No panic path, no unbounded map — the instrument map is keyed by 13 `&'static str`s.

### Finding 1 — pool mutex held across host-adapter (GIL) calls

`crates/monty-pool/src/pool.rs:339`

```rust
for _ in 0..lock_ignore_poison(&self.state).idle.len() {
    self.count_termination("closed");
}
```

Rust extends the `for` scrutinee's temporaries over the whole loop body, so the `MutexGuard` is live across every `count_termination` — each of which reenters the host adapter and takes the Python GIL. I confirmed the scoping with a standalone program (`try_lock` fails on every iteration). This is the exact inversion the rest of the file goes out of its way to avoid ("never call into the host adapter under the lock", `pool.rs:224`; `record_stable`'s docstring calls it "the deadlock the pool avoids everywhere").

Trigger: dropping a `Pool` without `close()` while workers sit idle — a supported shutdown. Impact: the pool mutex is held for N × GIL-handoff, blocking every concurrent checkout/release in the process; and it is a real deadlock if a host's `record_metric` ever reaches back into the pool. Fix is the one cubic suggested (thread still unresolved): snapshot `idle.len()`, then loop.

### Finding 2 — `record_stable`'s retry loop is unbounded, and degrades under exactly the conditions that trigger it

`crates/monty-pool/src/telemetry/metrics.rs:340`

The loop re-records until the atomic stops changing, with no iteration cap, and each pass is another call into the host SDK. The sandbox controls the input rate: every external call flips the shared idle counter twice, so `for _ in range(n): host_call()` across the pool keeps it moving. Measured amplification of `monty.pool.workers.idle` writes (Python binding, counting adapter):

| workers | sink latency | suspensions | idle-gauge records | amplification |
|---|---|---|---|---|
| 8 | ~0 | 2,400 | 5,814 | 1.21× |
| 16 | 0.5 ms | 4,800 | 32,681 | 3.40× |

It's a positive-feedback shape — the slower the host's sink, the more retries, which loads the sink further — and 0.5 ms is not pessimistic: I measured 0.5–2.4 ms per bridge call whenever any host thread is CPU-busy in Python. Bounding the retries (2–3) keeps the self-correcting behaviour without the tail. Also raised by cubic (P3), unresolved.

### Finding 3 — metrics roughly triple sandbox-controlled host-callback traffic (documented class, new magnitude)

Every frame now records on the pool's runtime thread; a printed line costs two `record_metric` calls (`monty.wire.frame.bytes`, `monty.print.bytes`) on top of the existing per-print log. 20,000 × `print('x')` with a telemetry adapter installed produced 40,019 metric calls + 20,000 log calls. Wall time for that feed, Python binding:

| | no adapter | `main` + adapter | this branch + adapter |
|---|---|---|---|
| host thread idle | 0.067 s | 0.19 s | 0.23 s |
| one host thread running plain Python | 0.95 s | 124.8 s | 143.8 s |

The 130× cliff is GIL handoff latency, is pre-existing, and is already written down in `limitations/pool-architecture.md` (which this PR correctly updates to cover measurements). What's new is that the per-frame instruments scale with sandbox *output rate* rather than with turns — they're the only instruments that do. If you want a lever, sampling or a per-turn aggregate for `monty.wire.frame.bytes` / `monty.print.bytes` would remove the sandbox's ability to set the host callback rate, at no cost to the other twelve instruments.

### Checked and clean

- **Gauge arithmetic under concurrency**: `LiveWorkersGauge::record`'s swap-then-`fetch_add` telescopes correctly however the two pools interleave; `to_isize` saturates; `Shared.idle` increments and decrements are balanced through the single `take_pending` path (`AtomicUsize` can't underflow from a reachable state, and would saturate rather than wrap if it did).
- **JS bridge queue overflow → permanent telemetry disable**: I expected metrics to fill napi's 1024-slot threadsafe queue and trip the one-way `disabled` flag, silencing telemetry process-wide. Did not reproduce: 200,000 printed lines with the JS event loop deliberately blocked for 4 s queued 400,015 metric + 200,004 span events, host RSS stayed ~100 MB, and a subsequent session still produced spans. (Both bindings pass an unparented context when `captureContext` returns nothing, so there is no previously-free untraced path that metrics newly make expensive.)
- **`worker.rs` frame-length bookkeeping**: `send_buf.len() - 4` cannot underflow — `encode_framed_into` clears and writes the 4-byte prefix before the body, and errors return before the subtraction.
- **Reference/lifecycle counting in `checkout.rs`**: `started` is taken by whichever of `finish`/`Drop` runs first, so a session is recorded once; `close()` drains `idle` before `Drop` counts it, so terminations aren't double-counted.
- No `unsafe`, no new allocation path that bypasses `ResourceTracker` (nothing sandbox-sized is built host-side; `close_suspension`'s 2–3 element attribute `Vec` is the only per-event allocation), no new `unwrap`/`expect` reachable from sandboxed input.

Unrelated trap I hit, in case it bites anyone else: `crates/monty-js/ts/binary.ts` prefers the installed `@pydantic/monty-<platform>` npm package over the workspace build, so after `make install-js` the JS tests run a *published* `monty` binary against the locally built addon and fail with `ChildEvent.kind: invalid wire type: Varint`. `MONTY_BIN=<workspace>/target/debug/monty` fixes it. Not a branch issue, but it looks alarmingly like a protocol desync.

**Comment by samuelcolvin** [comment]
@davidhewitt I need your help on this once you're back.

File: crates/monty-pool/src/telemetry/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,1345 @@
...
+use logfire::{ExponentialHistogram, Logfire};
+use monty_proto::{pb, pb::os_call::Call};
+use opentelemetry::{
+    KeyValue,
+    metrics::{Counter, UpDownCounter},
+};
+
+/// Live workers, whatever they are doing.
+static LIVE_WORKERS: Instrument = Instrument {
+    kind: MetricKind::UpDownCounter,
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply crates/monty-pool/src/telemetry/metrics.rs:37]: Sure but that's only if the implementation is wrong. With a correct implementation the up/down counters are a better choice semantically here.

File: crates/monty-pool/src/telemetry/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,1427 @@
...
+        let total = event.total_execution_micros;
+        let delta = total.saturating_sub(self.reported_micros);
+        self.reported_micros = self.reported_micros.max(total);
+        self.metrics
+            .record(&RUN_EXECUTION, MetricValue::seconds(Duration::from_micros(delta)), &[]);
+    }
+
+    /// Ends a terminal event as a run or housekeeping turn.
+    fn end_terminal(&mut self, outcome: &'static str, event: &pb::ChildEvent) {
+        if self.run_active {
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply crates/monty-pool/src/telemetry/metrics.rs:645]: When a `Dump` interrupts an active feed and the worker then terminates, two operations end:

- the `Dump` housekeeping turn
- the underlying feed, because its worker is gone

Current `end_terminal` calls `end_run`, which first calls `end_turn`. It therefore records both:

- `monty.turn.duration` with `fatal_error`/`shutdown`
- `monty.run.duration` with `fatal_error`/`shutdown`

The run is not recorded as completed, and its duration correctly includes suspension/housekeeping time per the metric definition. The execution delta also belongs to the active feed.

The suggested change would close only the `Dump`, then silently discard the active feed when `TurnMetrics` is dropped—the omission the previous review comment asked us to fix. No code change is needed.
